File: validate-binary-search-tree/validate-binary-search-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    TreeNode prevNode;
    boolean isValid = true;
    public boolean isValidBST(TreeNode root) {
        inorder(root);
        return isValid;
    }

    private void inorder(TreeNode node) {
        if (node == null) {
            return;
        }
        inorder(node.left);
        if (prevNode != null) {
            if (prevNode.val >= node.val) {
                isValid = false;
                return;
            }
        }
        prevNode = node;
        inorder(node.right);
    }
}

File: 98-validate-binary-search-tree/validate-binary-search-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def isValidBST(self, root: Optional[TreeNode]) -> bool:
        def helper(root):
            if not root:
                return (float('inf'), -float('inf'), True)
            leftMin, leftMax, leftValid = helper(root.left)
            rightMin, rightMax, rightValid = helper(root.right)
            myMin = min(leftMin, rightMin, root.val)
            myMax = max(leftMax, rightMax, root.val)
            myValid = leftMax < root.val and root.val < rightMin and leftValid and rightValid
            return (myMin, myMax, myValid)
        return helper(root)[2]